### DISCUSSION TITLE
Howto remove the merge commit (diamond) using lazygit

### DISCUSSION THREAD

**Discussion by suntsu42** [jesseduffield/lazygit#3454] [Q&A] [answered]
I'm starting with the following git log
![lg_log](https://github.com/jesseduffield/lazygit/assets/1768336/1001518d-db4a-4549-989b-8b368d86c657)

Goal is to remove the merge commit, but keep all actual commits. 
![lg_log_after_rebase](https://github.com/jesseduffield/lazygit/assets/1768336/0f66a67d-e120-4a4a-8ae7-2a541f57426d)

Using the console, i'd do 
```bash
 git rebase 20a10 feature/testbranc
``` 

I can't figure out howto do this operation in lazygit. 
I've tried interactive rebase and setting a base commit, but was unable to get the desired result. 

How do i get rid of the merge commit using lazygit?

**Comment by stefanhaller** [comment, marked as answer]
This is a very good question, and I'm afraid the answer is no, there is no way in lazygit to do this currently. The reason is that lazygit insists on adding `--rebase-merges` to every rebase command. (Which is a good thing, I don't think we should change that.)

If we do #2849 in one way or another, this could help, it would allow users to toggle the `--rebase-merges` case by case. But then, if users are knowledgeable enough to know that's what they need to do, they might just as well drop down to the command line and do it there.

Or is this a situation that you encounter frequently?

**Reply by tjwelde** [reply]
Could you elaborate on why it is a good thing to always add `--rebase-merges`? 

I e.g. have the use-case that a colleague merged main into their branch while working on a feature, but we want a clean git history. I took over the feature and want to clean it up by rebasing on main and getting rid of the merge commit. (Of course I can do it via the cmdline, but would love to do it in lazygit directly)

**Reply by stefanhaller** [reply]
What you describe is the one special case where you don't want `--rebase-merges`. I'd claim that in basically all other cases where you rebase a branch (and it contains merge commits that you made intentionally), you do want it. So it's better to err on the side of preserving them, I'd say.

I'm almost wondering if we should provide a separate command like "flatten commit history" or some such for the situation you described. It feels like something different from a rebase to me.

**Reply by matt-atticus** [reply]
is there a reason not to allow dropping merge commits during interactive rebase? I often work with stacked PRs/branches, which can accumulate merges commits (from merging in `main`) over time. When I get around to rebasing the end of the stack onto `main`, because I can't drop merge commits during the interactive rebase, my commit history ends up like:
- feat: some-feat-d
- merge branch `main` into `feat-c`
- merge branch `main` into `feat-b`
- merge branch `main` into `feat-a`

which is harmless since those merges are effectively empty commits at this stage, but its still clutter in the PR and the branch. Arguably I could avoid this by rebasing instead of merging at the time, but its often easier to merge because you can deal with conflicts in one go, rather than for every commit that's replayed during a rebase.

**Reply by stefanhaller** [reply]
> is there a reason not to allow dropping merge commits during interactive rebase?

Yes, and it is explained in the description of #4094.

There are several workarounds:
- run `git rebase --no-rebase-merges main` on the command line (or in lazygit's `:` shell prompt)
- make a custom command (e.g. `flatten`) that calls this
- drop the merge commits outside of an interactive rebase; this works

**Reply by matt-atticus** [reply]
understood - I hadn't realised I could drop merge commits directly, that's definitely easy enough! Thanks!

**Comment by suntsu42** [comment]
Fair enough. I just thought that i missed something. 
I'm new to lazygit and tried all the things i normally do in the console. 

> Or is this a situation that you encounter frequently?

No. I only need this when i made a merge commit by mistake, and it's a simple enough fix in the console.

**Comment by duzun** [comment]
I've just discovered `lazygit` today and also thought that I've missed something...

I wanted to remove a lot of merge commits, so I used a hybrid approach:
- in lazygit, find the commit to start with and ctrl+o to copy its hash,
- in CLI: `git rebase -i <hash>`
- continue solving the conflicts using lazygit
